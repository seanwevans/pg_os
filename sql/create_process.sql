-- tests for create_process
SET client_min_messages TO warning;
\i :abs_srcdir/../usersrolespermissions.sql
\set VERBOSITY terse

-- processes table and simplified create_process procedure
CREATE TABLE processes (
    id SERIAL PRIMARY KEY,
    name TEXT UNIQUE NOT NULL,
    state TEXT NOT NULL,
    priority INTEGER DEFAULT 1,
    created_at TIMESTAMP DEFAULT now(),
    updated_at TIMESTAMP DEFAULT now(),
    owner_user_id INTEGER REFERENCES users(id),
    duration INTEGER DEFAULT 1
);

CREATE OR REPLACE PROCEDURE create_process(
    process_name TEXT,
    owner_id INTEGER,
    process_priority INTEGER DEFAULT 1
)
LANGUAGE plpgsql
AS $$
BEGIN
    INSERT INTO processes (name, state, priority, owner_user_id, duration)
    VALUES (process_name, 'new', process_priority, owner_id, 1);
EXCEPTION
    WHEN unique_violation THEN
        RAISE EXCEPTION 'Process name % already exists', process_name;
END;
$$;

-- setup: create user
SELECT create_user('carol') AS user_id;

-- successful process creation
CALL create_process('proc_ok', 2, 5);
SELECT name, state, priority, owner_user_id FROM processes;

\set ON_ERROR_STOP off
-- duplicate process name should fail
CALL create_process('proc_ok', 2, 5);
-- invalid owner should fail
CALL create_process('other', 999, 1);
\set ON_ERROR_STOP on
