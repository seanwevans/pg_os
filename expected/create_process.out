-- tests for create_process
\set ECHO none
SELECT create_user('carol') AS user_id;
 user_id 
---------
       2
(1 row)

CALL create_process('proc_ok', 2, 5);
SELECT name, state, priority, owner_user_id FROM processes;
  name   | state | priority | owner_user_id 
---------+-------+----------+---------------
 proc_ok | new   |        5 |             2
(1 row)

CALL create_process('proc_ok', 2, 5);
ERROR:  Process name proc_ok already exists
CALL create_process('other', 999, 1);
ERROR:  insert or update on table "processes" violates foreign key constraint "processes_owner_user_id_fkey"
